Allow user to send template variable query string as raw JSON

IntegersOfK opened the pull request:
This is my way of addressing the issue I raised #155.

Edit: The strategy of this Pull Request initially started with a check to see if the string was actually JSON, but after discussion we've moved to a Switch next to the query editor where the user can choose if they want it sent as raw JSON or to continue using the new behaviour. The rest of this initial comment is no longer relevant.

<del>I will not be offended if you feel this update is too radical, but honestly I think this is the way the "JSON" plugin should handle it. If I put JSON into a query string for a template variable, I would expect it to be sent as actual JSON and not an escaped string in a key called target. It directly fits the spirit of the plugin's expected usage.</del>

<del>Otherwise, for non-valid JSON or regular strings, it still works the same way as before with `{"target":"querystring"}`. Admittedly, anyone who is already dealing with the "double escaping" issue for their JSON as a string in their query templates, will have to make a change, but I can tell you absolutely that they will be thankful for this fix. If you are okay to move forward with this approach, I would suggest something like the following release notes:</del>

> <del>New! If your Template Variable query string is valid JSON, it will now be sent to your backed as an actual JSON object in the request, rather than an escaped string. All regular string or non-valid JSON will maintain the existing behavior `{"target":"querystring"}` like before, and no changes are required.</del>
> 
><del> If you were already using what is considered valid JSON in a Template Variable query string, you will need to update your backend as you will no longer have to unescape the values from a string placed in a "target" key.</del>

<del>Given the way the tests are kind of checking for the response instead of the request body, I'm not sure exactly how to implement a test for this. I'm open to suggestions or taking another look at the tests though, I just wanted to get this pull request in for reference. </del>

<img width="1040" alt="2" src="https://user-images.githubusercontent.com/529833/111839025-8b811600-88d0-11eb-935e-c326e54284a4.PNG">
<img width="974" alt="1" src="https://user-images.githubusercontent.com/529833/111839026-8b811600-88d0-11eb-8222-02a9d94f314d.PNG">

IntegersOfK committed:
Parse and consider template variable query string as JSON when valid

simPod commented:
At the first sight this looks solid. Will check the code tomorrow or on Monday. Thanks for HQ PR!

simPod commented on README.md:
```diff
@@ -57,12 +57,22 @@ Those two urls are optional:
 Grafana issues this request on 
 
-1. _Variables > New/Edit_ page. `Query` field value is passed in a body as 
+1. _Variables > New/Edit_ page. `Query` field value is passed in a body as below (template variables are expanded as regex by default)
```
```suggestion
1. _Variables > New/Edit_ page. `Query` field value is passed in a body as shown below (template variables are expanded as regex by default)
```

simPod commented on src/DataSource.ts:
```diff
@@ -69,10 +69,15 @@ export class DataSource extends DataSourceApi<GrafanaQuery, GenericOptions> {
+    let interpolated;
+    try {
+      interpolated = JSON.parse(getTemplateSrv().replace(query, undefined, 'json'));
+    } catch (e) {
```
we should check for specific error here, not catch all

IntegersOfK replied:
Yeah I was thinking about how best to deal with this, because it's not really to catch an error so much as it is a check to see if the inputted query string is actual JSON or not (because we still want to be backwards compatible with people who want to send the legacy `{"target":"querystringhere"}` format I assume). I looked it up and apparently this really is the accepted way of checking if a string is valid JSON. I didn't love the idea of a try/catch myself either though so I understand your initial hesitation there.

It seems for the template variable query strings, we don't have the ability to modify the entry field to add something like a checkbox or anything resembling the "Additional JSON data" section like custom queries (which can come with their own query editor for that section that we can assume is JSON). Otherwise I would say we could put a checkbox there for the user to say, "This is a json string" and handle it explicitly rather than inferring based on the JSON validity of the string.

We could consider adding an option in the datasource configuration to say, "I use JSON in my template query strings" and then we could look at that setting in the function before we try to parse the string as JSON. Would you prefer that? The downside to that would be that it's one or the other; you're either using JSON in the template queries or you're using the traditional target/string approach. And then you're sort of at the whim of the person who configured the datasource to choose that for everyone who might Editors, but that's probably not a big concern considering the different queries have to be handled on the datasource backend anyway.

I digress, to me, the way I've implemented it here is the easiest and I do sort of feel like the try/catch fits the spirit of the plugin: If you are using JSON, we _try_ to send it to the backend datasource as that JSON. Otherwise, we _catch_ that you are not using valid JSON, and we force it into JSON by plopping it into `{"target":"yourquerystring"}`.

What do you think? Either way, we have to check for JSON or let the user tell us that it is JSON, or make a breaking change by telling everyone that they _must_ provide JSON here.

IntegersOfK replied:
Oh my mistake, there is a [VariableQueryEditor](https://grafana.com/docs/grafana/latest/developers/plugins/add-support-for-variables/#add-support-for-query-variables-to-your-data-source). Ok let me take a look at how that works and maybe we can make the option to use JSON or not more explicit.

IntegersOfK replied:
Ok I was able to get a first draft of the toggle switch working in a new VariableQueryEditor component!

![image](https://user-images.githubusercontent.com/529833/111931488-ad75c680-8a91-11eb-9383-e69bc4083287.png)

There is still an issue with toggling it on/off and I'm not sure about the new type yet, so let me try all the other features and see about making tests.

simPod commented on README.md:
````diff
@@ -57,12 +57,22 @@ Those two urls are optional:
 
+Alternatively, if you provide a full valid JSON string in the query field, it will be passed in the body as a native JSON object, and template variables will be expanded as a JSON array. For example, selecting two items in a multi-field dropdown `$myservers`
+```string
+{"selectedservers":$myservers}
````
can we add a test? What happens when no value/one entry is selected?

IntegersOfK replied:
It uses the default behavior of the Grafana interpolation for JSON. When one is selected, it provides it as a string:
![image](https://user-images.githubusercontent.com/529833/111911693-f3ed0600-8a3c-11eb-994e-16daf62e39b6.png)

And when nothing is selected, it leaves the string empty:
![image](https://user-images.githubusercontent.com/529833/111911660-dae45500-8a3c-11eb-90ac-f1ad08b98c70.png)

I also want to add a test, I'm just having trouble figuring out how that works for the input of this field. I will take another look.

IntegersOfK committed:
Update README.md

Co-authored-by: Simon Podlipsky <simon@podlipsky.net>

IntegersOfK committed:
Added new VariableQueryEditor component which allows the user to specify if they are using a normal string or JSON.

IntegersOfK committed:
Apply Prettier suggestions.

IntegersOfK commented:
Made some good progress here, it's definitely going to work it has a switch to toggle the behavior!

Ok, so the only thing left to do now is fix the existing tests, add new tests for the feature, and deal a bug where you can flick the switch on but not off if there's a parse error. Also I'm not sure how the whole empty string autocomplete thing is supposed to work and I have to learn about the "type" in metricFindQuery to make sure it all still works properly.

IntegersOfK committed:
Fix which allows the template var raw json option switch to be flicked off.

IntegersOfK committed:
Added test for and updated existing tests to accept new VariableQuery type instead of plain string.

IntegersOfK commented:
Ok I think it all works! I'll give it one more try with my backend tomorrow to be sure.

IntegersOfK commented:
At the moment, updating the plugin from existing queries, does not work (the query field is not populated). So I think that's the final issue to solved before we can move forward. I will keep working on it, thanks.

simPod reviewed (commented):
I wonder since we change UI can you replace screenshots in readme?

IntegersOfK commented:
> I wonder since we change UI can you replace screenshots in readme?

Sure I can definitely add a screenshot to the README. It's literally just one extra little switch when editing a template variable, so it shouldn't be too busy for the overview.

IntegersOfK committed:
Add legacy template variable query support.

IntegersOfK committed:
Add check for template var json validity, including when using other template vars in the object.

IntegersOfK committed:
Added screenshot of new raw json query switch and changed existing README images to relative refs.

IntegersOfK committed:
Run prettier properly.

IntegersOfK commented:
Alright, I'd say we're finally ready to do a real review of this!

- It supports the legacy format now too so existing dashboards will not break.
- I added a valid json check when the switch is toggled on, so the user knows that they've inputted proper JSON in the query string.
- I included a screenshot [as you requested](https://github.com/simPod/GrafanaJsonDatasource/pull/156#pullrequestreview-617574625) (and I changed the image links to relative in the README so they will work on any clone/fork).

I have a few thousand dashboards in production which use this data source, so tomorrow I will take a couple hours to see about using this new version to be sure it's all good with the legacy schema/existing dashboards as well as the new use case. Will report back on that soon.

simPod commented:
> and I changed the image links to relative in the README so they will work on any clone/fork

grafana requires it to be absolute because they render it on their site

> I took the initiative to update the version number

it will be done during release proces by npm so we can revert

simPod reviewed (changes requested):
This looks solid, gj! Will test it in my prod as well

IntegersOfK commented on src/DataSource.ts:
```diff
@@ -68,11 +69,23 @@ export class DataSource extends DataSourceApi<GrafanaQuery, GenericOptions> {
+    if (typeof legacyOrNew === 'object') {
+      query = legacyOrNew;
+    } else {
+      query = { query: legacyOrNew, asJson: false };
```
Ok first of all, these are all great points you're making and I'm totally happy to change approaches. I've put a lot of thought on how to deal with this so let me explain and you can tell me what you think we can do about it...

The purpose of this section is specifically to reconcile the new style (an actual object where we can put keys/values for this feature and anything else we would want in the future) with the existing dashboard schemas that everyone currently has stored in a regular string with this query field.

The existing schema:
```
...
`"templating": {
    "list": [
      {
        "allValue": null,
        "current": {},
        "datasource": null,
        "definition": "{\"Thisisatest\":\"test\"}",
        "description": null,
        "error": null,
        "hide": 0,
        "includeAll": false,
        "label": null,
        "multi": false,
        "name": "query0",
        "options": [],
        "query": "{\"Thisisatest\":\"test\"}",
        "refresh": 0,
        "regex": "",
        "skipUrlSync": false,
        "sort": 0,
        "tagValuesQuery": "",
        "tags": [],
        "tagsQuery": "",
        "type": "query",
        "useTags": false
      }
    ]
  }...
```

vs the new (which I have now changed `asJson` to `format`):
```
...
"templating": {
    "list": [
      {
        "allValue": null,
        "current": {},
        "datasource": null,
        "definition": "{\"Thisisatest\":\"test\"} (JSON)",
        "description": null,
        "error": null,
        "hide": 0,
        "includeAll": false,
        "label": null,
        "multi": false,
        "name": "query0",
        "options": [],
        "query": {
          "format": "json",
          "query": "{\"Thisisatest\":\"test\"}",
        },
        "refresh": 0,
        "regex": "",
        "skipUrlSync": false,
        "sort": 0,
        "tagValuesQuery": "",
        "tags": [],
        "tagsQuery": "",
        "type": "query",
        "useTags": false
      }...
```

If we don't have some way to convert the old "string" style to an actual object, everyone will have to re-create their template variable queries. By checking if we've received a plain string, we can convert it to the new. If they make any change to their template query and then resave their dashboard, it will be saved in the new format.

The downside is this annoying check, and it'll probably be years before we can assume everyone has touched their template variable queries at least once to remove the check. I don't know what else to do other than to piss everyone off by nulling out their existing dashboard template variable query strings when they update.

On the plus side, in the future it'll be way easier to expand this query object, because it's non-breaking to add another key (for example, maybe someday we will we want to add support to include the date range).

I did think about ways we could keep the traditional old string style and still add this new support, like maybe prepend the string with some sort of extra data to signify the type, but I think this really limits us and doesn't feel nicer than that check. Ie. we could make something that we can parse out of the string:

```
        "query": "ASJSON: {\"Thisisatest\":\"test\"}",
```

But then that still requires some sort of check to see if it's old or new, and it's back to the same problem as before where we're parsing the string to see if it's valid JSON or not at query time. So I figured we might as well just switch to an actual type (VariableQuery) and suck up the check to help people upgrade. In a future release maybe we can tell everyone that they must have updated their template variable queries since this version and re-saved their dashboards at least once, to remove that check.

simPod replied:
Ah right. Can we keep it simple, so the code would stay as is so people can migrate their dashboards.

And will drop the condition in the next "minor" version (because we're still at `0.x.y`).

IntegersOfK replied:
Yes I think that's good, or maybe we could drop it in the first release of this plugin after Grafana 8.0 comes out this year. I find at those times people have a high tolerance for potentially breaking changes if you don't follow proper upgrade path (ie. most people who use the UI we can just tell them that they must flick the new JSON switch back and forth once and re-save their dashboards).

simPod replied:
Can you please prepare PR after this gets merged with the drop?

IntegersOfK commented:
> > and I changed the image links to relative in the README so they will work on any clone/fork
> 
> grafana requires it to be absolute because they render it on their site
> 
> > I took the initiative to update the version number
> 
> it will be done during release proces by npm so we can revert

Ok I didn't know, thanks for explaining I reverted all that. The new screenshot should work after merge.

IntegersOfK committed:
Update src/VariableQueryEditor.tsx

Co-authored-by: Simon Podlipsky <simon@podlipsky.net>

IntegersOfK committed:
Change asJSON from a boolean to a string so it can be expanded to different formats in the future.

IntegersOfK committed:
Minor fix to equal check.

IntegersOfK commented on src/VariableQueryEditor.tsx:
```diff
@@ -0,0 +1,69 @@
+    }
+  };
+
+  function checkValidJSON(query: VariableQuery) {
```
Whoops, Visual Studio Code did it that way when I chose "declare new function..." I have corrected it thanks.

IntegersOfK commented on src/types.d.ts:
```diff
@@ -23,6 +23,11 @@ export interface GrafanaQuery extends DataQuery {
 export interface GenericOptions extends DataSourceJsonData {}
 
+export interface VariableQuery {
+  query: any;
```
Indeed it is, I have corrected it.

IntegersOfK commented on src/QueryEditor.tsx:
```diff
@@ -23,6 +23,11 @@ interface LastQuery {
 
+const blankVar: VariableQuery = {
+  query: '',
+  asJson: false,
```
Good idea, I like that because maybe people will want csv or who knows what in the future. I had to add an extra function to deal with the boolean nature of the switch but that's not going to break anything in the future if it changes!

IntegersOfK committed:
Updated test for change to format (rather than isJson).

IntegersOfK commented:
Tried this on about a dozen dashboards today, with no issues so far!

simPod commented on spec/datasource.jest.ts:
```diff
@@ -165,7 +165,7 @@ describe('GenericDatasource', () => {
     setTemplateSrv(templateSrvStub);
 
-    ds.metricFindQuery('search', Format.Timeseries).then((result) => {
+    ds.metricFindQuery({query:'search', asJson: false}, Format.Timeseries).then((result) => {
```
the type checker is off for spec, asjson is no longer there

IntegersOfK replied:
Whoops, thought I already did that in my previous commit but I think I missed it because these tests pass considering at part of the object isn't really considered in the test. Thanks for catching that.

simPod commented on src/types.d.ts:
```diff
@@ -23,6 +23,11 @@ export interface GrafanaQuery extends DataQuery {
 
+export interface VariableQuery {
+  query: string;
+  format: string;
```
can we enumerate options through string union?

```suggestion
  format: 'string' | 'json';
```

also I wonder whether to use `'string'` or some better name 🤔

IntegersOfK replied:
> can we enumerate options through string union?

Oh cool, I didn't even know we could do that. Is there something else we should update in our checks for string vs json?

> also I wonder whether to use `'string'` or some better name 🤔

Yeah I think we talked about "raw" but after looking at some of the other datasources I feel like the concept of "raw" implies that it is the base representation of any format type. For example, "raw json" can be conflated with "raw query" when really they sort of imply different things, neither of which is the fact that the type of entry people are using in this case is in fact a string. Therefore, I went with "string" because that's what it is, a plain old query **string**.

I don't feel strongly about that though, it's whatever you want to to... I'm just happy we're going to be able to supply the template variables as JSON. 🥇

IntegersOfK committed:
Updated expected type of tests to new format.

IntegersOfK committed:
Update src/types.d.ts

Co-authored-by: Simon Podlipsky <simon@podlipsky.net>

simPod reviewed (approved):
I guess rebase and 🚢 

Thanks!

KnutZuidema commented:
This would be a great addition for our use case of this plugin. Is there anything preventing you (@simPod) from rebasing this and merging?

simPod committed:
Convert query data to DataFrame

IntegersOfK commented:
Hi, apologies if I'm a bit unclear @simPod on what we should do next. Am I supposed to "rebase" or will you take care of that step? I'm happy to do anything we need to get a version with this feature cut and it seems like quite a few others are hopeful to see it soon too. Thanks!

simPod commented:
@IntegersOfK yes, rebase please. I have changed the base to v0.2 since I've started working on v0.3 on master.

IntegersOfK committed:
Reconcile rebase changes with existing work.

IntegersOfK commented:
Ok I rebased and have tested it on my own instance, seems to be working as expected.

I think the only issue will be the images I added to the README because they do not exist at a static location until you've merged them into your repo, at which point I think we can just update it again to use the new URLs provided by github.

simPod commented:
I seems I have no permissions to push to your branch. Can you remove 2cac0339a783a0a723a8afcd5d7a161f08a44db5 commit please? It is supposed to be on v0.3 only

simPod commented on README.md:
```diff
@@ -32,7 +32,7 @@ See [here](https://grafana.com/grafana/plugins/simpod-json-datasource/) for more
 When adding datasource add your API endpoint to the `URL` field. That's where datasource will make requests to.
 
-![Datasource setup](https://raw.githubusercontent.com/simPod/grafana-json-datasource/master/docs/images/datasource-setup.png)
+![Datasource setup](./docs/images/datasource-setup.png)
```
```suggestion
![Datasource setup](https://raw.githubusercontent.com/simPod/grafana-json-datasource/v0.2.x/docs/images/datasource-setup.png)
```

IntegersOfK committed:
Update README.md

Co-authored-by: Simon Podlipsky <simon@podlipsky.net>

IntegersOfK committed:
Revert Dataframes changes until they are ready for next version

IntegersOfK commented:
Ok I made you a collaborator on my fork, feel free to modify/rewrite/rebase/squash as you see fit. And sorry about the confusion/branch mess, when this all started it was supposed to be a quick little parse check I threw on the master branch but now it turned into a full fledge optional feature with legacy checks, etc.

simPod commented:
@IntegersOfK I tried to push to master but github does not authenticate me still 
![image](https://user-images.githubusercontent.com/327717/122664477-ec570d80-d1a1-11eb-9234-b1d7ce13ae09.png)

In the meantime, I'm preparing last changes.

simPod commented:
thank you! 94ee7d8b47030c1c87ba3e1b064fd9e50ad6d778

simPod commented:
@IntegersOfK I'm planning to issue a 0.2.x release. Have you had a chance to already run the [0.2.x branch](https://github.com/simPod/GrafanaJsonDatasource/tree/v0.2.x)?
